### ISSUE TITLE
Is C++ required?

### ISSUE THREAD

**Issue by Josef-Vonasek** [luna/Dataframes#4]
Since Luna can do pointer arithmetic and should be pretty fast on its own, why not implement the library in pure Luna, without using C?

**Comment by mwu-tow** [comment]
TLDR: in the longer-term we are going to migrate C++ parts to Luna.

Long answer:  
I think that C++ parts could be expressed in Luna right now but it wouldn't work very well.

Luna is still a work-in-progress in some areas. In particular performance and optimization capabilities are still being actively developed — Luna is not "that fast" yet. When it comes to DataFrames, being able to quickly process huge amounts of data is crucial. As things stand right now, it makes sense to implement low-level primitives and operations on them in C++ and to consume them from Luna.

Such approach will allow Luna users to use Dataframes to quickly process and visualize their data. For many users this will be enough — being able to use in Luna libraries like Dataframes and being able to implement them in pure Luna are separate use-cases, and since we can deliver the first one, we don't want to wait for the second one. At this point we prioritize data-processing and visualization capabilities of Luna over being able to implement low-level code (i.e. rivalling C and C++).

Implementing Dataframes needs much more than just having support for pointer arithmetics. Providing highly performant primitives also requires control over data layout in memory. Many defaults that Luna currently assumes (like lazy evaluation, boxed values, garbage collecting, being interpreted rather than compiled, etc.) are often not best suited for implementing low-level code. That's common case for many high-level languages and is also the reason why C and C++ remain still so prevalent. They are just tools perfectly suited for the low-level domain. I don't think that taking advantage of them is wrong — that's what's Luna's FFI support is for. 

This library is still pretty much at the design space exploration stage. The goal is to deliver well-performing low-level pieces and see how they perform, and work on integration with Luna Studio (e.g. data visualizations).  Having the initial implementation in C++ will provide a good benchmark for us. In the long term we want to move away from C++ and use Luna as much as reasonably possible. My belief is that it is best to first write the performant code (be it C++ or anything) and then gradually move it to Luna while preserving the speed. That will also allow us to catch design and architectural issues early.
